### ISSUE TITLE
Detect unnecessary cast statements

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#228]
The `cast` function should be used sparingly because it completely disables type checking for that expression. In cases where the cast is a no-op, there's no good reason to keep it. The type checker should provide an option to detect these no-op cast calls and flag them as errors.

**Comment by erictraut** [comment]
This will be in the next version.

**Comment by erictraut** [comment]
This is now implemented in 1.0.59, which I just published.

**Comment by higorc-mck** [comment]
I see an issue here. I am working with Spark, which has the `toPandas()` method in their DataFrames which are annotated to return `DataFrameLike`, which are _not_ annotated the same as `pandas.DataFrame` at all. As I'm trying to cast from `DataFrameLike` to `DataFrame` I'm getting this error, where I believe it clearly shouldn't. The situations are clearly unequivalent, as I can actually get rid of many errors downstream by doing this `cast` call.

Here's an excerpt:

```python
import pandas as pd
from pyspark.sql import DataFrame

...

def test_add_ids(users: DataFrame) -> None:
    users_with_ids = add_ids(users)
    pd_users_with_ids = cast(pd.DataFrame, users_with_ids.toPandas())

    assert pd_users_with_ids.columns == [SomeEnum.SOURCE, SomeEnum.ROW] + users.columns # I get a pyright error here without cast, says `columns` is of type `Column`
    assert pd_users_with_ids.shape[0] == 6 # same error here
```

**Comment by erictraut** [comment]
@higorc-mck, this issue has been closed for four years. If you would like to report a bug, please open a new issue and include a minimal, self-contained code sample. Your sample above references symbols that are not defined.

If you are not sure about the intended behavior, please start a thread in the Discussion tab.
